TypeScript program codegen: We use the `?` operator instead of the `?.` operator.

<!-- Please provide a general summary of the issue, including what you're trying to accomplish. -->
We generate typescript code that makes invalid use of the `?` operator. We should be using the `?.` operator. See here: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html.

Specifically, we should see the expression `loggings?[0]?.targetBucket` change to `loggings?.[0]?.targetBucket`.
## Steps to reproduce
<!-- Provide a link to a live example or an unambiguous set of steps to reproduce this bug. Tell us what you expected to happen, and what happened instead. Please include code to reproduce, if relevant. -->

1. `cd pkg/codegen/internal/test/testdata/aws-s3-logging-pp`
2. `tsc --noEmit aws-s3-logging.ts`


Expected: Clean compile
Actual:
```sh
aws-s3-logging.ts(8,89): error TS1005: ':' expected.
```

--- a/changelog/pending/20221128--programgen-nodejs--add-between-and.yaml
+++ b/changelog/pending/20221128--programgen-nodejs--add-between-and.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: programgen/nodejs
+  description: Add `.` between `?` and `[`.

--- a/pkg/codegen/nodejs/gen_program_expressions.go
+++ b/pkg/codegen/nodejs/gen_program_expressions.go
@@ -594,8 +594,20 @@ func (g *generator) genRelativeTraversal(w io.Writer, traversal hcl.Traversal, p
 			contract.Failf("unexpected traversal part of type %T (%v)", part, part.SourceRange())
 		}
 
+		var indexPrefix string
 		if model.IsOptionalType(model.GetTraversableType(parts[i])) {
 			g.Fgen(w, "?")
+			// `expr?[expr]` is not valid typescript, since it looks like a ternary
+			// operator.
+			//
+			// Typescript solves this by inserting a `.` in before the `[`: `expr?.[expr]`
+			//
+			// We need to do the same when generating index based expressions.
+			indexPrefix = "."
+		}
+
+		genIndex := func(inner string, value interface{}) {
+			g.Fgenf(w, "%s["+inner+"]", indexPrefix, value)
 		}
 
 		switch key.Type() {
@@ -604,13 +616,13 @@ func (g *generator) genRelativeTraversal(w io.Writer, traversal hcl.Traversal, p
 			if isLegalIdentifier(keyVal) {
 				g.Fgenf(w, ".%s", keyVal)
 			} else {
-				g.Fgenf(w, "[%q]", keyVal)
+				genIndex("%q", keyVal)
 			}
 		case cty.Number:
 			idx, _ := key.AsBigFloat().Int64()
-			g.Fgenf(w, "[%d]", idx)
+			genIndex("%d", idx)
 		default:
-			g.Fgenf(w, "[%q]", key.AsString())
+			genIndex("%q", key.AsString())
 		}
 	}
 }

--- a/pkg/codegen/testing/test/program_driver.go
+++ b/pkg/codegen/testing/test/program_driver.go
@@ -88,7 +88,7 @@ var PulumiPulumiProgramTests = []ProgramTest{
 	{
 		Directory:   "aws-s3-logging",
 		Description: "AWS S3 with logging",
-		SkipCompile: allProgLanguages.Except("python").Except("dotnet"),
+		SkipCompile: codegen.NewStringSet("go"),
 		// Blocked on nodejs: TODO[pulumi/pulumi#8068]
 		// Flaky in go: TODO[pulumi/pulumi#8123]
 	},

--- a/pkg/codegen/testing/test/testdata/aws-s3-logging-pp/nodejs/aws-s3-logging.ts
+++ b/pkg/codegen/testing/test/testdata/aws-s3-logging-pp/nodejs/aws-s3-logging.ts
@@ -5,4 +5,4 @@ const logs = new aws.s3.Bucket("logs", {});
 const bucket = new aws.s3.Bucket("bucket", {loggings: [{
     targetBucket: logs.bucket,
 }]});
-export const targetBucket = bucket.loggings.apply(loggings => loggings?[0]?.targetBucket);
+export const targetBucket = bucket.loggings.apply(loggings => loggings?.[0]?.targetBucket);

